pkg/middleware/headers.go

```diff
@@ -43,7 +43,7 @@ func newStripHeaders(headers []options.Header) alice.Constructor {
 
 func flattenHeaders(headers http.Header) {
 	for name, values := range headers {
-		if len(values) > 1 {
+		if len(values) > 1 && name != "Set-Cookie" {
```

Comment:
Would like to suggest a comment here just to add a link to the mozilla doc mentioned in the issue
```suggestion
		// Set-Cookie should not be flattened, ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
		if len(values) > 1 && name != "Set-Cookie" {
```